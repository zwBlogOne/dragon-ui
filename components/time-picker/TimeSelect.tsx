import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import rafObj from '../utils/rAF';
import Select from '../select';
import TimeSelectProps from './PropsType';

const scrollTo = (element, to, duration) => {
  if (duration <= 0) {
    element.scrollTop = to;
    return;
  }

  const difference = to - element.scrollTop;
  const perTick = (difference / duration) * 10;

  rafObj.rAF(() => {
    element.scrollTop += perTick;

    if (element.scrollTop === to) { return; }
    scrollTo(element, to, duration - 10);
  });
};

class TimeSelect extends Component<TimeSelectProps, any> {
  private hourDom: string;
  private minuteDom: string;
  private secondDom: string;

  constructor (props) {
    super(props);
    this.state = {
      selectH: [],
      selectedH: [],

      selectM: [],
      selectedM: [],

      selectS: [],
      selectedS: [],
    };
  }

  componentDidMount () {
    this.selectInit();
    this.setDataFromValue(this.props);
  }

  componentWillReceiveProps (nextProps) {
    this.setDataFromValue(nextProps);
  }

  setDataFromValue (props) {
    const arrTime = props.value.split(':');

    this.setState({
      selectedH: [arrTime[0]],
      selectedM: [arrTime[1]],
      selectedS: [arrTime[2]],
    });
  }

  scrollToSelected (element, selectIndex, duration) {
    // @ts-ignore
    const selectDom: any = ReactDOM.findDOMNode(element).children[0];
    const list = selectDom.querySelector('ul');
    const topOption = list.children[selectIndex];
    const to = topOption.offsetTop;

    if (!list) { return; }
    scrollTo(selectDom, to, duration);
  }

  selectInit () {
    const { selectH, selectM, selectS } = this.state;

    for (let i = 0; i < 60; i++) {
      const num = i < 10 ? '0' + i : i + '';
      if (i < 24) { selectH.push(num); }
      selectM.push(num);
      selectS.push(num);
    }
    this.setState({ selectH, selectM, selectS });
  }

  onSelect (type, element, arr, selectIndex) {
    let { selectedH, selectedM, selectedS } = this.state;
    if (type === 'H') { selectedH = arr; }
    if (type === 'M') { selectedM = arr; }
    if (type === 'S') { selectedS = arr; }
    this.setState({ selectedH, selectedM, selectedS });
    const rVal = selectedH[0] + ':' + selectedM[0] + ':' + selectedS[0];
    this.scrollToSelected(element, selectIndex, 120);
    this.props.onChange(rVal);
  }

  render () {
    const { selectH, selectM, selectS } = this.state;

    return (
      <div className="ui-time-select">
        <Select.Multiple
          value={this.state.selectedH}
          ref={el => this.hourDom = el}
          onChange={(_, row) => this.onSelect('H', this.hourDom, [row.value], row.index)}
        >
          {
            selectH && selectH.map((option, index) => {
              return <Select.Option key={index} value={option}>{option}</Select.Option>;
            })
          }
        </Select.Multiple>
        <Select.Multiple
          value={this.state.selectedM}
          ref={el => this.minuteDom = el}
          onChange={(_, row) => this.onSelect('M', this.minuteDom, [row.value], row.index)}
        >
          {
            selectM && selectM.map((option, index) => {
              return <Select.Option key={index} value={option}>{option}</Select.Option>;
            })
          }
        </Select.Multiple>
        <Select.Multiple
          value={this.state.selectedS}
          ref={el => this.secondDom = el}
          onChange={(_, row) => this.onSelect('S', this.secondDom, [row.value], row.index)}
        >
          {
            selectS && selectS.map((option, index) => {
              return <Select.Option key={index} value={option}>{option}</Select.Option>;
            })
          }
        </Select.Multiple>
      </div>
    );
  }
}

export default TimeSelect;
