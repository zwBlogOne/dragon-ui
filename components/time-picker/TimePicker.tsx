import React, { Component, Fragment } from 'react';
import classnames from 'classnames';
import Dropdown from '../dropdown';
import Icon from '../icon';
import TimeSelect from './TimeSelect';
import LocaleReceiver from '../locale/LocaleReceiver';

type placement = 'bottomLeft' | 'bottomCenter' | 'bottomRight' | 'topLeft' | 'topCenter' | 'topRight';

export interface TimePickerProps {
  value?: any;
  defaultValue?: string;
  placeholder?: string;
  isDisabled?: boolean;
  isRadius?: string;
  size?: any;
  style?: React.CSSProperties;
  locale: { clear: string, confirm: string };
  localeCode: string;
  dropdownStyle?: React.CSSProperties;
  placement?: placement;
  onChange?: (value: any) => void;
}

class TimePicker extends Component<TimePickerProps, any> {
  static defaultProps = {
    onChange: () => {},
  };

  private unmounted: boolean;
  private dropdownEl: React.ReactNode;

  constructor (props) {
    super(props);
    const value = props.value || props.defaultValue;
    this.unmounted = false;
    this.state = {
      value,
      dropdown: false,
    };
  }

  componentDidMount () {
    this.unmounted = true;
  }

  componentWillReceiveProps (nextProps: TimePickerProps) {
    if ('value' in nextProps) {
      this.setState({ value: nextProps.value });
    }
  }

  onDateChange (value) {
    if (value.indexOf('undefined') > -1) { return; }
    this.setState({
      value,
    }, () => {
      this.setDropdown(!!value, this.props.onChange);
    });
  }

  setDropdown (isOpen, callback?) {
    if (!this.unmounted) { return; }

    this.setState({
      dropdown: isOpen,
    }, () => {
      if (callback) {
        callback(this.state.value);
      }
    });
  }

  isNodeInTree (node, tree) {
    while (node) {
      if (node === tree) {
        return true;
      }
      node = node.parentNode;
    }

    return false;
  }

  onVisibleChange = visible => {
    this.setState({
      dropdown: visible,
    });
  }

  onConfirmBtn () {
    this.setDropdown(false, this.props.onChange);
  }

  render () {
    const {
      defaultValue, placeholder, isDisabled, isRadius, size, locale,
      localeCode, dropdownStyle, placement, onChange, ...others
    } = this.props;
    const { value, dropdown } = this.state;
    const disabled = 'disabled' in this.props || isDisabled;
    const radius = 'radius' in this.props || isRadius;

    let valueText = placeholder || '00:00:00';
    let hasValue = false;

    if (value) {
      valueText = value;
      hasValue = true;
    }

    const cls = classnames({
      'ui-select': true,
      'ui-time-select-wrapper': true,
      'ui-select-open': dropdown,
      disabled,
      radius,
      [`size-${size}`]: !!size,
    });
    const textCls = classnames({
      'ui-select-text': true,
      'ui-select-text-placeholder': !hasValue,
    });

    const overlay = (
      <Fragment>
        <TimeSelect
          value={valueText}
          onChange={(value) => this.onDateChange(value)}
        />
        <div className="ui-select-bottom">
          <a className="clear-btn" href="javascript:;" onClick={() => this.onDateChange('')}>
            {locale.clear}
          </a>
          <a className="confirm-btn" href="javascript:;" onClick={() => this.onConfirmBtn()}>
            {locale.confirm}
          </a>
        </div>
      </Fragment>
    );

    return (
      <div className={cls} style={this.props.style} {...others}>
        <Dropdown
          ref={el => this.dropdownEl = el}
          visible={dropdown}
          disabled={disabled}
          style={{ width: 240, ...dropdownStyle }}
          zIndex={2020}
          onVisibleChange={this.onVisibleChange}
          overlay={overlay}
          placement={placement}
        >
          <div
            className="ui-select-selection"
            role="combobox"
            aria-autocomplete="list"
            aria-haspopup="true"
            aria-expanded="false"
          >
            <span className={textCls}>{valueText}</span>
            <Icon className="ui-time-select-icon" type="time-circle"/>
          </div>
        </Dropdown>
      </div>
    );
  }
}

export default LocaleReceiver(TimePicker, 'TimePicker');
