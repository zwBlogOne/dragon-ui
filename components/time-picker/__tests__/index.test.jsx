import React from 'react';
import ReactDOM from 'react-dom';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import TimePicker from '../index';
import TimeSelect from '../TimeSelect';

// ReactDOM.createPortal = node => node

describe('TimePicker', () => {
  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element) => {
      return element;
    });
  });

  afterEach(() => {
    ReactDOM.createPortal.mockClear();
  });

  it('renders normal time-picker correctly', () => {
    const wrapper = render(
      <div>
        <TimePicker />
      </div>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders disabled time-picker correctly', () => {
    const wrapper = render(
      <div>
        <TimePicker
          isDisabled
        />
      </div>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders disabled time-select correctly', () => {
    const wrapper = render(
      <TimeSelect
        value="00:00:00"
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders isRadius time-picker correctly', () => {
    const wrapper = render(
      <div>
        <TimePicker
          isRadius
          style={{ width: 160 }}
        />
      </div>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('behaves correctly when item click', () => {
    const onChange = jest.fn();
    const initValue = '00:00:00';
    const wrapper = mount(
      <TimeSelect
        value={initValue}
        onChange={onChange}
      />
    );
    wrapper.find('.ui-option-list').first().simulate('click');

    expect(onChange).toHaveBeenCalledWith(initValue);
  });

  it('behaves correctly when clear button click', () => {
    const wrapper = mount(<TimePicker value="12:33:22" />);
    wrapper.find('.clear-btn').simulate('click');

    expect(wrapper.find('.ui-select-text').text()).toBe('00:00:00');
  });
});

