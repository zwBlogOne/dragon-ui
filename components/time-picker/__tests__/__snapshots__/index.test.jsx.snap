// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker renders disabled time-picker correctly 1`] = `
<div>
  <div
    class="ui-select ui-time-select-wrapper disabled"
  >
    <div
      class="ui-dropdown-trigger-box"
    >
      <div
        aria-autocomplete="list"
        aria-expanded="false"
        aria-haspopup="true"
        class="ui-select-selection"
        role="combobox"
      >
        <span
          class="ui-select-text ui-select-text-placeholder"
        >
          00:00:00
        </span>
        <i
          class="ui-icon ui-icon-time-circle theme-default ui-time-select-icon"
        />
      </div>
    </div>
    <div
      class="ui-dropdown"
      style="min-width:0;width:240px;left:0;top:0;position:absolute;animation-duration:300ms;display:none;overflow:hidden;z-index:2020"
    >
      <div
        class="ui-time-select"
      >
        <span
          class="ui-select"
        >
          <span
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            class="ui-select-selection"
            role="combobox"
            style="height:100%;max-height:250px;overflow:auto"
          >
            <ul
              class="ui-menu ui-menu-light ui-menu-inline"
              role="menu"
            />
          </span>
        </span>
        <span
          class="ui-select"
        >
          <span
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            class="ui-select-selection"
            role="combobox"
            style="height:100%;max-height:250px;overflow:auto"
          >
            <ul
              class="ui-menu ui-menu-light ui-menu-inline"
              role="menu"
            />
          </span>
        </span>
        <span
          class="ui-select"
        >
          <span
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            class="ui-select-selection"
            role="combobox"
            style="height:100%;max-height:250px;overflow:auto"
          >
            <ul
              class="ui-menu ui-menu-light ui-menu-inline"
              role="menu"
            />
          </span>
        </span>
      </div>
      <div
        class="ui-select-bottom"
      >
        <a
          class="clear-btn"
          href="javascript:;"
        >
          清除
        </a>
        <a
          class="confirm-btn"
          href="javascript:;"
        >
          确认
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`TimePicker renders disabled time-select correctly 1`] = `
<div
  class="ui-time-select"
>
  <span
    class="ui-select"
  >
    <span
      aria-autocomplete="list"
      aria-expanded="false"
      aria-haspopup="true"
      class="ui-select-selection"
      role="combobox"
      style="height:100%;max-height:250px;overflow:auto"
    >
      <ul
        class="ui-menu ui-menu-light ui-menu-inline"
        role="menu"
      />
    </span>
  </span>
  <span
    class="ui-select"
  >
    <span
      aria-autocomplete="list"
      aria-expanded="false"
      aria-haspopup="true"
      class="ui-select-selection"
      role="combobox"
      style="height:100%;max-height:250px;overflow:auto"
    >
      <ul
        class="ui-menu ui-menu-light ui-menu-inline"
        role="menu"
      />
    </span>
  </span>
  <span
    class="ui-select"
  >
    <span
      aria-autocomplete="list"
      aria-expanded="false"
      aria-haspopup="true"
      class="ui-select-selection"
      role="combobox"
      style="height:100%;max-height:250px;overflow:auto"
    >
      <ul
        class="ui-menu ui-menu-light ui-menu-inline"
        role="menu"
      />
    </span>
  </span>
</div>
`;

exports[`TimePicker renders isRadius time-picker correctly 1`] = `
<div>
  <div
    class="ui-select ui-time-select-wrapper radius"
    style="width:160px"
  >
    <div
      class="ui-dropdown-trigger-box"
    >
      <div
        aria-autocomplete="list"
        aria-expanded="false"
        aria-haspopup="true"
        class="ui-select-selection"
        role="combobox"
      >
        <span
          class="ui-select-text ui-select-text-placeholder"
        >
          00:00:00
        </span>
        <i
          class="ui-icon ui-icon-time-circle theme-default ui-time-select-icon"
        />
      </div>
    </div>
    <div
      class="ui-dropdown"
      style="min-width:0;width:240px;left:0;top:0;position:absolute;animation-duration:300ms;display:none;overflow:hidden;z-index:2020"
    >
      <div
        class="ui-time-select"
      >
        <span
          class="ui-select"
        >
          <span
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            class="ui-select-selection"
            role="combobox"
            style="height:100%;max-height:250px;overflow:auto"
          >
            <ul
              class="ui-menu ui-menu-light ui-menu-inline"
              role="menu"
            />
          </span>
        </span>
        <span
          class="ui-select"
        >
          <span
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            class="ui-select-selection"
            role="combobox"
            style="height:100%;max-height:250px;overflow:auto"
          >
            <ul
              class="ui-menu ui-menu-light ui-menu-inline"
              role="menu"
            />
          </span>
        </span>
        <span
          class="ui-select"
        >
          <span
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            class="ui-select-selection"
            role="combobox"
            style="height:100%;max-height:250px;overflow:auto"
          >
            <ul
              class="ui-menu ui-menu-light ui-menu-inline"
              role="menu"
            />
          </span>
        </span>
      </div>
      <div
        class="ui-select-bottom"
      >
        <a
          class="clear-btn"
          href="javascript:;"
        >
          清除
        </a>
        <a
          class="confirm-btn"
          href="javascript:;"
        >
          确认
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`TimePicker renders normal time-picker correctly 1`] = `
<div>
  <div
    class="ui-select ui-time-select-wrapper"
  >
    <div
      class="ui-dropdown-trigger-box"
    >
      <div
        aria-autocomplete="list"
        aria-expanded="false"
        aria-haspopup="true"
        class="ui-select-selection"
        role="combobox"
      >
        <span
          class="ui-select-text ui-select-text-placeholder"
        >
          00:00:00
        </span>
        <i
          class="ui-icon ui-icon-time-circle theme-default ui-time-select-icon"
        />
      </div>
    </div>
    <div
      class="ui-dropdown"
      style="min-width:0;width:240px;left:0;top:0;position:absolute;animation-duration:300ms;display:none;overflow:hidden;z-index:2020"
    >
      <div
        class="ui-time-select"
      >
        <span
          class="ui-select"
        >
          <span
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            class="ui-select-selection"
            role="combobox"
            style="height:100%;max-height:250px;overflow:auto"
          >
            <ul
              class="ui-menu ui-menu-light ui-menu-inline"
              role="menu"
            />
          </span>
        </span>
        <span
          class="ui-select"
        >
          <span
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            class="ui-select-selection"
            role="combobox"
            style="height:100%;max-height:250px;overflow:auto"
          >
            <ul
              class="ui-menu ui-menu-light ui-menu-inline"
              role="menu"
            />
          </span>
        </span>
        <span
          class="ui-select"
        >
          <span
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            class="ui-select-selection"
            role="combobox"
            style="height:100%;max-height:250px;overflow:auto"
          >
            <ul
              class="ui-menu ui-menu-light ui-menu-inline"
              role="menu"
            />
          </span>
        </span>
      </div>
      <div
        class="ui-select-bottom"
      >
        <a
          class="clear-btn"
          href="javascript:;"
        >
          清除
        </a>
        <a
          class="confirm-btn"
          href="javascript:;"
        >
          确认
        </a>
      </div>
    </div>
  </div>
</div>
`;
